target/linux/generic/pending-5.10/920-mangle_bootargs.patch

```diff
@@ -31,19 +31,27 @@ Signed-off-by: Imre Kaloz <kaloz@openwrt.org>
  	help
 --- a/init/main.c
 +++ b/init/main.c
-@@ -608,6 +608,29 @@ static inline void setup_nr_cpu_ids(void
+@@ -608,6 +608,37 @@ static inline void setup_nr_cpu_ids(void
  static inline void smp_prepare_cpus(unsigned int maxcpus) { }
  #endif
  
 +#ifdef CONFIG_MANGLE_BOOTARGS
 +static void __init mangle_bootargs(char *command_line)
 +{
-+	char *rootdev;
++	char *rootdev = NULL;
++	char *next_rootdev = command_line;
 +	char *rootfs;
 +
-+	rootdev = strstr(command_line, "root=/dev/mtdblock");
++	/*
++	 * Find the last bootargs, so buggy bootargs can be masked with
++	 * "bootargs-append" DT property.
++	 */
++	while (next_rootdev = strstr(next_rootdev, "root=")) {
++		rootdev = next_rootdev;
++		next_rootdev += 5;
++	}
 +
-+	if (rootdev)
++	if (rootdev && !strncmp(rootdev, "root=/dev/mtdblock", 18)
```

Comment:
sorry if I already asked... But why this additional check?

Reply:
so u can mask old (buggy) one completely. In my case uboot always passes root=/dev/mtdblock0 which would of course lead to a unbootable system, so I could mask it with something like root=/nonexistance and let mtdsplit decide the correct root.

Reply:
Ok now I'm confused...

Can you by chance give me the output of the bootargs passed to a running system (exactly what the kernel print from the bootlog) so I can understand this more clearly?

And what should be expected?

The strncmp is fragile since other format can be used and that would result in mangling only the format that match /dev/mtdblock

Reply:
Same logic applies to `rootfstype` => `mangled_fs`

Reply:
Ok let me clear some thing and tell me if i'm wrong....

Wtih bootargs-append the bootargs would result in something like this right? (your bootargs are taken from a running system or are ""simulated""?)

`Kernel command line: console=ttyS0,115200 mtdparts=nand_mtd:0xc0000@0(uboot)ro,0x7f00000@0x100000(root) rootfstype=squashfs root=/dev/mtdblock3 rw mv_net_config=(00:11:88:0f:62:81,0:1:2:3),mtu=1500 cur_runtime_image_num=2 ro root=`
(notice at the end ro root=)

With the current code (master branch without this patch applied)
Bootargs would result in:
`Kernel command line: console=ttyS0,115200 mtdparts=nand_mtd:0xc0000@0(uboot)ro,0x7f00000@0x100000(root) mangled_fs=squashfs mangled_rootblock=3 rw mv_net_config=(00:11:88:0f:62:81,0:1:2:3),mtu=1500 cur_runtime_image_num=2 ro root=`

And with the current pr it will result in:
`Kernel command line: console=ttyS0,115200 mtdparts=nand_mtd:0xc0000@0(uboot)ro,0x7f00000@0x100000(root) rootfstype=squashfs root=/dev/mtdblock3 rw mv_net_config=(00:11:88:0f:62:81,0:1:2:3),mtu=1500 cur_runtime_image_num=2 ro root=`
effectively not mangling anything


So you final target is using bootargs-append to disable the mangle behavior?

Reply:
yes pretty like that (and that's from a real running system)

Reply:
if that is from a running system where is the appended stuff ?

Reply:
bootargs-append dts property

Reply:
Here is an example

https://github.com/openwrt/openwrt/blob/721206a1b14fc5eba7ba8e20d979b26baf934e0d/target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-nas1.dts#L19-L24

Reply:
@yangfl but why you say it's malfunctioning?

also the mangle patch is there from a lot of times and the nas1 dts was introduced later so it's either a malfunctioning device was pushed or we are getting confused here....

so the thing here is that the mangle patch works as expected and mangle the root and rootfstype from bootargs... and then we append an empty root= that is probably insert by a patch

What you want is to use what the bootloader provide and disable the mangle patch right?

If that's the case then I would just expand the mangle patch to add a special string and skip mangling the bootargs... 

something like 

`bootargs-append = " console=ttyS0,115200 disablemangle"; `

Hope I finally got this right.

If i'm right the solution is much simpler to both explain and implement than change how the current mangle patch behave and cause many regression on other target.

Reply:
It's probably nas1 passed some rootdev other than `root=/dev/mtdblock` so that this issue won't emerge on that device.

I don't want to use what the bootloader provide. In fact, I don't care original bootargs content, as long as only the last parameter takes effect.

So here is my point: the original mangle patch is doing something _against_ the kernel behavior. Instead of taking the last `root=` (see init/do_mounts.c#L297-L303), it may turn into other previous `root=` and cause confusing.

To me `disablemangle` might also solve the problem on my device, but it does not solve the fundamental voilation of kernel behavior; new option just hide it.

Same argument also applys on `rootfstype`.

Reply:
> It's probably nas1 passed some rootdev other than `root=/dev/mtdblock` so that this issue won't emerge on that device.

No if I got this right... we mangle the bootloader root value... then we use 202-linksys-find-active-root.patch to select the right partition provided by the bootloader... the passed empty root= is just to not mess with the kernel... 


> I don't want to use what the bootloader provide. In fact, I don't care original bootargs content, as long as only the last parameter takes effect.

So you want the bootloader root value to be mangled and pass an empty root= 

> So here is my point: the original mangle patch is doing something _against_ the kernel behavior. Instead of taking the last `root=` (see init/do_mounts.c#L297-L303), it may turn into other previous `root=` and cause confusing.

We can assume stuff here... From a bootloader view it won't make sense to pass multiple root= values in bootargs... So the mangle patch assume only one root=, mangle that and bootargs-append add an empty root=

> To me `disablemangle` might also solve the problem on my device, but it does not solve the fundamental voilation of kernel behavior; new option just hide it.

That was useful only if you needed to use the root from the bootloader but you don't want to do that.

So yet again I'm missing what is the final target.... (and sorry if I'm getting annoying but I really need to understand this and how to proceed)

I mean without this patch your device doesn't boot at all? The wrong partition is used? 

I mean what are the case of a bootloader passing 2 root= value ?

Reply:
> No if I got this right... we mangle the bootloader root value... then we use 202-linksys-find-active-root.patch to select the right partition provided by the bootloader... the passed empty root= is just to not mess with the kernel...

Hmm interesting.

> I mean without this patch your device doesn't boot at all? The wrong partition is used?

Yes, by 202-linksys-find-active-root.patch.

> I mean what are the case of a bootloader passing 2 root= value ?

OK I understand... I might be misleaded on the usage of this patch. I should take look at 202-linksys-find-active-root.patch where mangled_rootblock is used. Sorry and thank you.

Reply:
Yes ideally you should check how 202 works and try to make the dual boot work with that patch... if that doesn't work we can consider tweaking 202 to disable his behavior. That seems to be a better suited and safe solution.

Here to help you in understand this better since I'm also curious

Reply:
one solution might be generalize existing `openwrt,cmdline-match` and drop `MANGLE_BOOTARGS` completely, since `MANGLE_BOOTARGS` does thing in a nasty way and cannot be enabled by default. See my #10674

Reply:
Did you by chance find a way to to make the current implementation work ? Did you encounter some blocker?

Reply:
I can't figure out a neat way to conditionally enable 202-linksys-find-active-root.patch (maybe through a device tree property, but can't find a proper node to place it in). Nevertheless better to close this pr anyway.

Reply:
@yangfl the idea was to use it to support your device for dual partition... not to disable it. Is it that impossible?

Reply:
Sorry I can't get your point... but as for this pr, I think I did wrong on mangled_rootblock and it should be closed, though further discussion on dual boot for these devices may be conducted.